pond_test.go

```diff
@@ -44,5 +44,32 @@ func TestPurgeAfterPoolStopped(t *testing.T) {
 
 	// Simulate purger goroutine attempting to stop a worker after tasks channel is closed
 	atomic.StoreInt32(&pool.stopped, 1)
-	pool.stopIdleWorker()
+	pool.maybeStopIdleWorker()
+}
+
+// See: https://github.com/alitto/pond/issues/33
+func TestPurgeDuringSubmit(t *testing.T) {
+
+	pool := New(1, 1)
+
+	var doneCount int32
+
+	// Submit a task to ensure at least 1 worker is started
+	pool.SubmitAndWait(func() {
+		atomic.AddInt32(&doneCount, 1)
+	})
+
+	assertEqual(t, 1, pool.IdleWorkers())
+
+	// Stop an idle worker right before submitting another task
+	pool.maybeStopIdleWorker()
```

Comment:
Thanks for this PR and the prompt reaction!

How does this test validate issue #33, if `maybeStopIdleWorker` is synchronously called by the same go routine?

Reply:
This test was entering a deadlock quite consistently (although not 100% of the time). `maybeStopIdleWorker` executes synchronously up to [this point](https://github.com/alitto/pond/blob/master/pond.go#L402), where it sends a `nil` task to kill a worker goroutine asynchronously (buffered channel). In the previous version of the lib, the idle worker counter was being updated when the killed goroutine exited, so there was a small window of time during which [this call](https://github.com/alitto/pond/blob/master/pond.go#L412) would incorrectly return `false`, indicating there was an idle worker available.